Feature/0001 base structure

ajax-torbin-v opened the pull request.

Commit:
Added .gitignore file

Commit:
Added base models, service, repository & controllers. Covered code by test

ajax-malov-n commented on src/main/kotlin/com/example/delivery/controller/UserController.kt:
```diff
@@ -0,0 +1,34 @@
+        return userService.findById(id)
+    }
+
+    @PutMapping("/add-order")
```
'add' in request url

vetterocki commented on src/main/kotlin/com/example/delivery/controller/ProductController.kt:
```diff
@@ -0,0 +1,34 @@
+@RequestMapping("/api/products")
+class ProductController(private val productService: ProductService) {
+
+    @GetMapping("/findById/{id}")
```
I suggest remove unnecessary words in each endpoint mapping (like 'findById' in URI etc)

ajax-malov-n commented on src/main/kotlin/com/example/delivery/model/MongoOrder.kt:
```diff
@@ -0,0 +1,36 @@
+import org.springframework.data.annotation.TypeAlias
+import org.springframework.data.mongodb.core.mapping.Document
+
+@TypeAlias("Order")
```
Why don't you use just 'MongoOrder'?

ajax-samondros-b replied:
Using type aliases is a good practice since Mongo use full class name with package by default which makes package structure refactoring more complex. Always use type aliases 👍

ajax-malov-n commented on src/main/kotlin/com/example/delivery/repository/impl/OrderRepositoryImpl.kt:
```diff
@@ -0,0 +1,37 @@
+
+@Repository
+class OrderRepositoryImpl(var mongoTemplate: MongoTemplate) : OrderRepository {
+    private val className = MongoOrder::class.java
```
Separate variables and methods with a blank line

ajax-malov-n commented on src/main/kotlin/com/example/delivery/service/OrderService.kt:
```diff
@@ -0,0 +1,55 @@
+        return order.toDTO()
+    }
+
+    fun updateStatus(updateOrderDTO: UpdateOrderDTO) {
+        if (!orderRepository.existsById(updateOrderDTO.id))
+            throw NotFoundException("Order with id " + updateOrderDTO.id + "doesn't exists")
+        orderRepository.updateOrderStatus(updateOrderDTO.id, updateOrderDTO.status)
+    }
+
+    fun deleteById(id: String) =
+        if (orderRepository.existsById(id)) orderRepository.deleteById(id)
+        else throw  NotFoundException("Order with id $id doesn't exists")
+}
```
Use the same logic for handling exceptions

ajax-malov-n commented on src/test/kotlin/com/example/delivery/controller/UserControllerTests.kt:
```diff
@@ -0,0 +1,87 @@
+import org.springframework.boot.test.mock.mockito.MockBean
+import org.springframework.http.MediaType
+import org.springframework.test.web.servlet.MockMvc
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
```
Remove * from imports

ajax-malov-n commented on src/test/kotlin/com/example/delivery/service/OrderServiceTests.kt:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Test
+    @DisplayName("Find existing order")
```
Remove @Display you can convey the information with fun 'name'

ajax-malov-n reviewed (commented):
Nice, but look through the comments

Commit:
Resolved issues. Refactored code

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/DeliveryApplication.kt:
```diff
@@ -0,0 +1,11 @@
+class DeliveryApplication
+
+fun main(args: Array<String>) {
+	runApplication<DeliveryApplication>(*args)
```
```suggestion
    runApplication<DeliveryApplication>(*args)
```

ajax-gnatiuk-s replied:
4 spaces instead of tab

ajax-gnatiuk-s commented on build.gradle.kts:
```diff
@@ -0,0 +1,46 @@
+	kotlin("plugin.spring") version "1.9.24"
+	id("org.springframework.boot") version "3.3.2"
+	id("io.spring.dependency-management") version "1.1.6"
+	kotlin("plugin.jpa") version "1.9.24"
```
```suggestion
    kotlin("plugin.jpa") version "1.9.24"
```
Pls, fix the whole file

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/config/MongoConfig.kt:
```diff
@@ -0,0 +1,22 @@
+        return "delivery"
+    }
+
+    override fun mongoClient() = MongoClients.create("mongodb://localhost:27017")
```
the database name and the connection url should be read from properties. Hardcoding is bad practice

ajax-samondros-b replied:
move url to application.yaml

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/exception/NotFoundException.kt:
```diff
@@ -0,0 +1,4 @@
+package com.example.delivery.exception
+
+class NotFoundException(message: String) : RuntimeException(message) {
+}
```
```suggestion
class NotFoundException(message: String) : RuntimeException(message)
```

ajax-gnatiuk-s commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,7 @@
+spring.data.mongodb.port=27017
+spring.data.mongodb.database=delivery
+spring.data.mongodb.username=slava
+spring.data.mongodb.password=12345
```
```suggestion
spring.data.mongodb.password=12345

```

See how to configure your IDE https://github.com/ajax-tarnavskyi-m/BookExchangeService/pull/1#discussion_r1718917610

ajax-gnatiuk-s replied:
It's better to use yaml instead of properties

ajax-gnatiuk-s commented on src/test/kotlin/com/example/delivery/TestObjects.kt:
```diff
@@ -0,0 +1,70 @@
+    password = "password",
+    orderIds = mutableListOf("1"),
+)
+val userDTO = user.toDTO()
```
such code should be placed in [test fixtures](https://docs.gradle.org/current/userguide/java_testing.html#sec:java_test_fixtures)

You can ask for more details after the daily

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/controller/OrderController.kt:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @PutMapping
+    fun update(@RequestBody updateOrderDTO: UpdateOrderDTO): OrderDTO {
```
Should we have path param `@PathVariable id: String` for the update operation?

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/controller/ProductController.kt:
```diff
@@ -0,0 +1,42 @@
+        return productService.findById(id)
+    }
+
+    @GetMapping("/all")
```
```suggestion
    @GetMapping
```

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/model/MongoOrder.kt:
```diff
@@ -0,0 +1,36 @@
+        val building: String = "none",
+        val index: String = "none") {
+    }
+}
```
fix according to https://kotlinlang.org/docs/coding-conventions.html#interface-implementation-layout

ajax-gnatiuk-s commented on src/test/kotlin/com/example/delivery/service/OrderServiceTests.kt:
```diff
@@ -0,0 +1,77 @@
+        Mockito.`when`(orderRepository.updateOrderStatus(any(), any())).thenReturn(updatedOrder)
+
+        orderService.updateStatus(UpdateOrderDTO("1", MongoOrder.Status.CANCELED))
+        verify(orderRepository, times(1)).updateOrderStatus("1", MongoOrder.Status.CANCELED)
```
```suggestion
        // GIVEN
        val updatedOrder = order.copy(status = MongoOrder.Status.CANCELED)
        Mockito.`when`(orderRepository.updateOrderStatus(any(), any())).thenReturn(updatedOrder)

        // WHEN
        orderService.updateStatus(UpdateOrderDTO("1", MongoOrder.Status.CANCELED))
        
        // THEN
        verify(orderRepository, times(1)).updateOrderStatus("1", MongoOrder.Status.CANCELED)
```

The suggested comments make your tests more readable

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/config/MongoConfig.kt:
```diff
@@ -0,0 +1,22 @@
+class MongoConfig : AbstractMongoClientConfiguration() {
+
+    override fun getDatabaseName(): String {
+        return "delivery"
```
move to application.yaml

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/dto/request/CreateOrderDTO.kt:
```diff
@@ -0,0 +1,10 @@
+
+data class CreateOrderDTO(
+    val items: Map<String, Int>,
+    val shipmentDetails: MongoOrder.ShipmentDetails
```
Do not use Mongo classes in your DTO

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/dto/request/UpdateOrderDTO.kt:
```diff
@@ -0,0 +1,9 @@
+
+data class UpdateOrderDTO (
+    val id: String,
+    val status: MongoOrder.Status
```
do not reference Mongo classes and its children

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/dto/response/OrderDTO.kt:
```diff
@@ -0,0 +1,12 @@
+    val id: String,
+    val items: Map<String, Int>, //TODO: Use List<Product> instead, fetched from a db
+    val totalPrice: Double,
+    val shipmentDetails: MongoOrder.ShipmentDetails,
```
do not reference Mongo classes and its children

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoOrder.kt:
```diff
@@ -0,0 +1,36 @@
+@TypeAlias("Order")
+@Document(collection = MongoOrder.COLLECTION_NAME)
+data class MongoOrder(
+    @Id val id: String? = null,
```
Use ObjectId as Id always. Do not build any logic around this Id. If you want to provide any identifier to the customer create separate field for it

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoOrder.kt:
```diff
@@ -0,0 +1,36 @@
+data class MongoOrder(
+    @Id val id: String? = null,
+    val items: Map<String, Int>?,
+    val totalPrice: Double?,
```
Use BigDecimal for prices

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoOrder.kt:
```diff
@@ -0,0 +1,36 @@
+    val shipmentDetails: ShipmentDetails?,
+    val status: Status? = Status.NEW) {
+    enum class Status {
+        NEW, SHIPMENT, COMPLETED, CANCELED, UNKNOWN
```
remove UNKNOWN since null can act as it. Replace `SHIPMENT` by `SHIPPING`

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoOrder.kt:
```diff
@@ -0,0 +1,36 @@
+        const val COLLECTION_NAME = "order"
+    }
+
+    fun toDTO(): OrderDTO = OrderDTO(
```
IMHO, You overloading this Mongo class by redundant logic. Make mappers instead of trying map inside data class. Same is true where you mapping to Mongo in your DTO

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoOrder.kt:
```diff
@@ -0,0 +1,36 @@
+        status ?: Status.UNKNOWN)
+
+    data class ShipmentDetails(
+        val city: String = "none",
```
every Mongo object should have nullable fields always. This is an embedded Mongo Object as Well. Call it as "MongoShipmentDetails". Do not forget about Type Alias

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoProduct.kt:
```diff
@@ -0,0 +1,27 @@
+    val amountAvailable: Int?,
+    val measurement: String?) {
+
+    fun toDTO(): ProductDTO = ProductDTO(
```
same about mapping

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoUser.kt:
```diff
@@ -0,0 +1,27 @@
+    val password: String?,
+    val orderIds: MutableList<String>?) {
+
+    fun toDTO(): UserDTO = UserDTO(
```
same about mapping

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoUser.kt:
```diff
@@ -0,0 +1,27 @@
+@TypeAlias("User")
+@Document(collection = MongoUser.COLLECTION_NAME)
+data class MongoUser (
+    @Id val id: String?,
```
ObjectId

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoProduct.kt:
```diff
@@ -0,0 +1,27 @@
+@TypeAlias("Product")
+@Document(collection = MongoProduct.COLLECTION_NAME)
+data class MongoProduct(
+    @Id val id: String? = null,
```
ObjectId

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoProduct.kt:
```diff
@@ -0,0 +1,27 @@
+@Document(collection = MongoProduct.COLLECTION_NAME)
+data class MongoProduct(
+    @Id val id: String? = null,
+    val name: String?,
```
It is a good practice to declare all fields with default null value

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/repository/OrderRepository.kt:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.stereotype.Repository
+
+@Repository
+interface OrderRepository  {
```
add findAll

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/repository/impl/OrderRepositoryImpl.kt:
```diff
@@ -0,0 +1,41 @@
+
+    override fun updateOrderStatus(id: String, status: MongoOrder.Status): MongoOrder?{
+        val query = Query.query(Criteria.where("_id").`is`(id))
+        val update = Update.update("status", status)
```
Use property reference instead of string. For example `MongoOrder::status.name`

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/repository/impl/OrderRepositoryImpl.kt:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    override fun deleteById(id: String) {
+        val query = Query.query(Criteria.where("_id").`is`(id))
```
use isEqualTo to avoid ` escaping

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/repository/impl/ProductRepositoryImpl.kt:
```diff
@@ -0,0 +1,37 @@
+        mongoTemplate.remove(query)
+    }
+
+    override fun findByName(name: String): MongoProduct? {
```
Think about index here

ajax-gnatiuk-s replied:
later, when we start mongo topic

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/model/MongoUser.kt:
```diff
@@ -0,0 +1,27 @@
+    val fullName: String?,
+    val phone: String?,
+    val password: String?,
+    val orderIds: MutableList<String>?) {
```
remove it. add reference to the user in the MongoOrder instead

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/repository/impl/UserRepositoryImpl.kt:
```diff
@@ -0,0 +1,35 @@
+
+    override fun findById(id: String): MongoUser? = mongoTemplate.findById(id, className)
+
+    override fun addOrder(userId: String, orderId: String): MongoUser? {
```
redundant. see comments above

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/service/OrderService.kt:
```diff
@@ -0,0 +1,56 @@
+    fun add(createOrderDTO: CreateOrderDTO): OrderDTO {
+        var sum = 0.0
+        createOrderDTO.items.mapNotNull { (productId, amount) ->
+            val product: MongoProduct = productRepository.findById(productId)
```
- Convert MongoProduct to DomainProduct. Then Write separate ProductReserveService and do any checks there. 
- Merge "has no available amount" and "Not enough items" altogether.

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/service/OrderService.kt:
```diff
@@ -0,0 +1,56 @@
+            sum += price * amount
+            product to amount
+        }.toMap()
+        val order: MongoOrder = orderRepository.save(MongoOrder(
```
where your Product's amountAvailable changes?

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/service/ProductService.kt:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.stereotype.Service
+
+@Service
+class ProductService(private val productRepository: ProductRepository) {
```
Add method to reserve products (decrease available amount)

ajax-samondros-b commented on src/main/kotlin/com/example/delivery/service/UserService.kt:
```diff
@@ -0,0 +1,36 @@
+        return user.toDTO()
+    }
+
+    fun addOrder(userId: String, orderId: String): UserDTO {
```
redundant

ajax-samondros-b commented on src/test/kotlin/com/example/delivery/TestObjects.kt:
```diff
@@ -0,0 +1,70 @@
+import com.example.delivery.model.MongoOrder
+import com.example.delivery.model.MongoUser
+
+val createProductDTO = CreateProductDTO("Coca-cola", 24.50, 69, "0.5L")
```
Create ProductFixture, UserFixture, OrderFixture objects instead. Create methods inside them

ajax-samondros-b commented on src/test/kotlin/com/example/delivery/controller/OrderControllerTests.kt:
```diff
@@ -0,0 +1,109 @@
+
+@ExtendWith(MockitoExtension::class)
+@WebMvcTest(OrderController::class)
+class OrderControllerTests {
```
`OrderControllerTest`

ajax-samondros-b commented on src/test/kotlin/com/example/delivery/controller/OrderControllerTests.kt:
```diff
@@ -0,0 +1,109 @@
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+import org.mockito.Mockito
```
prefer MockK over Mockito

ajax-samondros-b commented on src/test/kotlin/com/example/delivery/controller/OrderControllerTests.kt:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Test
+    fun `should return order when order exists` () {
```
good practice to make

```
// GIVEN

// WHEN

// THEN
```

blocks

ajax-samondros-b commented on src/test/kotlin/com/example/delivery/service/OrderServiceTests.kt:
```diff
@@ -0,0 +1,77 @@
+    fun `should return order when order exists` () {
+        Mockito.`when`(orderRepository.findById("1")).thenReturn(order)
+
+        assertEquals(orderService.findById("1"), orderDTO)
```
prefer to create objects per test and not use shared objects

ajax-samondros-b commented on src/test/kotlin/com/example/delivery/service/ProductServiceTests.kt:
```diff
@@ -0,0 +1,66 @@
+import kotlin.test.assertEquals
+
+@ExtendWith(MockitoExtension::class)
+class ProductServiceTests {
```
make all test classes as internal

Commit:
Wrapped controller responses into ResponseEntity

ajax-torbin-v committed:
Removed dependencies from Mongo model
Moved mapping functions to separate class
Converted price to BigDecimal

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/controller/UserController.kt:
```diff
@@ -0,0 +1,45 @@
+import org.springframework.web.bind.annotation.RestController
+
+@RestController
+@RequestMapping("/api/users")
```
```suggestion
@RequestMapping("/users")
```

we can set the global context in config instead of specifying the prefix in each controller

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/controller/UserController.kt:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @GetMapping("/{id}")
+    fun findById(@PathVariable id: String): ResponseEntity<UserDTO> {
+        return ResponseEntity.ok(userService.findById(id))
```
```suggestion
    fun findById(@PathVariable id: String): UserDTO = userService.findById(id)
```
it does the same.

Is there any reason to return `ResponseEntity` instead of a DTO?

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/controller/UserController.kt:
```diff
@@ -0,0 +1,45 @@
+
+    @DeleteMapping("/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    fun deleteById(@PathVariable id: String): ResponseEntity<Void> {
+        userService.deleteById(id)
+        return ResponseEntity.status(HttpStatus.NO_CONTENT).build()
+    }
```
```suggestion
    fun deleteById(@PathVariable id: String) = userService.deleteById(id)
```

Response status could specified using the corresponding annotation

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/mapper/OrderMapper.kt:
```diff
@@ -0,0 +1,4 @@
+package com.example.delivery.mapper;
+
+public class OrderMapper {
```
should we remove it?

It's blank

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/repository/UserRepository.kt:
```diff
@@ -0,0 +1,13 @@
+    fun existsById(id: String): Boolean
+    fun save(user: MongoUser): MongoUser
+    fun findById(id: String): MongoUser?
+    fun addOrder(userId: String, orderId: String): MongoUser?
```
why it returns nullabe object?

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/service/UserService.kt:
```diff
@@ -0,0 +1,36 @@
+        return updatedUser.toDTO()
+    }
+
+    fun deleteById(id: String) =
+        if (userRepository.existsById(id)) userRepository.deleteById(id)
+        else throw NotFoundException("User with id $id doesn't exists")
```
let's make it more readable
```suggestion
    fun deleteById(id: String) = if (userRepository.existsById(id)) {
        userRepository.deleteById(id)
    } else {
        throw NotFoundException("User with id $id doesn't exists")
    }
```

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/service/ProductService.kt:
```diff
@@ -0,0 +1,25 @@
+
+    fun add(createProductDTO: CreateProductDTO): ProductDTO = productRepository.save(createProductDTO.toModel()).toDTO()
+
+
```
```suggestion
```
redundant empty line

ajax-gnatiuk-s commented on src/test/kotlin/com/example/delivery/controller/OrderControllerTests.kt:
```diff
@@ -0,0 +1,109 @@
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
+
+@ExtendWith(MockitoExtension::class)
```
```suggestion
```
It's spring test, we don't need this extension

ajax-gnatiuk-s commented on src/test/kotlin/com/example/delivery/controller/OrderControllerTests.kt:
```diff
@@ -0,0 +1,109 @@
+
+    @BeforeEach
+    fun setUp() {
+        MockitoAnnotations.openMocks(this)
```
why do we need this?

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/service/OrderService.kt:
```diff
@@ -0,0 +1,56 @@
+            shipmentDetails = createOrderDTO.shipmentDetails)
+        )
+        return order.toDTO()
+    }
```
let's refactor this function, it's too complex

ajax-torbin-v committed:
Refactored DTO`s
Added trailing comma.
Moved orderIds from UserDTO to userId in OrderDTO

ajax-torbin-v committed:
Refactored repository and controller layers
Removed addOrder from UserController

ajax-torbin-v committed:
Added Domain models
Moved Mongo models to separate package

ajax-torbin-v committed:
Refactored code to work with Domain model.
Moved business logic to ProductReserveService

ajax-torbin-v committed:
Refactored tests and controller layer.
Created Fixture objects.
Added given-when-then clause
Removed ResponseEntity from controllers
Moved UpdateDTO id field as @PathVariable

ajax-torbin-v committed:
Added logic to decrease amount of product when adding order

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/controller/OrderController.kt:
```diff
@@ -0,0 +1,45 @@
+        @PathVariable id: String,
+        @RequestBody updateOrderDTO: UpdateOrderDTO,
+    ): OrderDTO {
+        return orderService.updateStatus(id, updateOrderDTO).toDTO()
```
be consistent. Use either block body or expression body at least in a particular file

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/controller/UserController.kt:
```diff
@@ -0,0 +1,36 @@
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    fun deleteById(@PathVariable id: String) = userService.deleteById(id)
+
+
```
```suggestion
```

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/config/GlobalExceptionHandler.kt:
```diff
@@ -0,0 +1,25 @@
+    @ExceptionHandler(NotFoundException::class)
+    fun handleNotFoundException(ex: NotFoundException, request: WebRequest): ResponseEntity<ErrorDetails> {
+        val errorDetails = ErrorDetails(
+            message = ex.message ?: "Something went wrong",
```
```
class NotFoundException(message: String) : RuntimeException(message)
```
message is not nullable in the exception

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/service/ProductService.kt:
```diff
@@ -0,0 +1,30 @@
+class ProductService(private val productRepository: ProductRepository) {
+
+    fun findById(id: String): DomainProduct {
+        val product = productRepository.findById(id) ?: throw NotFoundException("Product with id $id not found")
```
`findById` - should return `null` if an entity wasn't found.
`getById` - should throw if an entity wasn't found.

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/service/ProductService.kt:
```diff
@@ -0,0 +1,30 @@
+        if (productRepository.existsById(id)) {
+            productRepository.deleteById(id)
+        } else {
+            throw NotFoundException("Product with id $id not found")
```
don't throw if entity doesn't exist for the delete operation

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/service/ProductReserveService.kt:
```diff
@@ -0,0 +1,30 @@
+    fun calculateTotalPrice(items: Map<DomainProduct, Int>): BigDecimal {
+        return items.entries.fold(BigDecimal.ZERO) { acc, (product, amount) ->
+            val price = product.price
+            if (price < BigDecimal.ZERO) throw ArithmeticException("Price of product ${product.id} is 0 or negative")
```
pls, fix the error message or change the condition to `price <= BigDecimal.ZERO`

ajax-gnatiuk-s replied:
Don't use `ArithmeticException` here, see my comment below

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/service/ProductReserveService.kt:
```diff
@@ -0,0 +1,30 @@
+    private fun validateProductAvailability(product: DomainProduct, amount: Int) {
+        val availableAmount = product.amountAvailable
+        if (availableAmount < amount) {
+            throw ArithmeticException("Not enough items for product ${product.id}")
```
Create a custom exception. 

From `ArithmeticException` javadocs:
```
 * Thrown when an exceptional arithmetic condition has occurred. For
 * example, an integer "divide by zero" throws an
 * instance of this class.
```

The exception is about arithmetic errors, not about business logic

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/service/ProductReserveService.kt:
```diff
@@ -0,0 +1,30 @@
+
+@Service
+class ProductReserveService {
+    fun reserveProducts(items: Map<DomainProduct, Int>) {
```
Suggestion to improve your code:

```kotlin
data class AccountedProduct(
    val product: DomainProduct,
    val amount: Int,
)

fun reserveProducts(products: List<AccountedProduct>) {
    // ...
}
```

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/mapper/OrderMapper.kt:
```diff
@@ -0,0 +1,49 @@
+
+    fun MongoOrder.toDomain(): DomainOrder = DomainOrder(
+        id,
+        items = items ?: mutableMapOf(),
```
```suggestion
        items = items ?: emptyMap(),
```
prefer immutable collections whenever it's possible

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/repository/OrderRepository.kt:
```diff
@@ -0,0 +1,14 @@
+
+@Repository
+interface OrderRepository {
+    fun existsById(id: String): Boolean
+    fun findById(id: String): MongoOrder?
+    fun save(order: MongoOrder): MongoOrder
+    fun updateOrderStatus(id: String, status: MongoOrder.Status): MongoOrder?
+    fun deleteById(id: String)
+    fun findAll(id: String): List<MongoOrder>
```
```suggestion
    fun existsById(id: String): Boolean
    fun findById(id: String): MongoOrder?
    fun findAll(id: String): List<MongoOrder>
    fun save(order: MongoOrder): MongoOrder
    fun updateOrderStatus(id: String, status: MongoOrder.Status): MongoOrder?
    fun deleteById(id: String)
```

ajax-gnatiuk-s replied:
move `findAll` closer to `findById`

ajax-gnatiuk-s commented on src/test/kotlin/com/example/delivery/controller/UserControllerTest.kt:
```diff
@@ -0,0 +1,72 @@
+        //GIVEN
+        doNothing().`when`(userService).deleteById("1")
+
+        //THEN
+        mockMvc.perform(delete("/users/{id}", "1"))
+            .andExpect(status().isNoContent)
+
+        verify(userService).deleteById("1")
```
```suggestion
        // WHEN // THEN
        mockMvc.perform(delete("/users/{id}", "1"))
            .andExpect(status().isNoContent)

        // AND THEN
        verify(userService).deleteById("1")
```

- `given` is setup block
- `when` is actual action to test
- `then` is assertion block

Sometimes `when` and `then` block shouldn't be splitted(as in your case here) so we can use `// WHEN // THEN` or `// WHEN-THEN` comment.

ajax-gnatiuk-s commented on src/test/kotlin/com/example/delivery/service/OrderServiceTest.kt:
```diff
@@ -0,0 +1,96 @@
+        //GIVEN
+        Mockito.`when`(orderRepository.findById("1")).thenReturn(null)
+
+        //THEN
```
```suggestion
        // WHEN //THEN
```

don't forget about the most important `when` comment. This test a bit confusing at first look, because it's not clear what we test here

ajax-gnatiuk-s commented on src/test/kotlin/com/example/delivery/service/OrderServiceTest.kt:
```diff
@@ -0,0 +1,96 @@
+
+    @Test
+    fun `should add order with proper dto`() {
+
```
```suggestion
```

ajax-gnatiuk-s commented on src/test/kotlin/com/example/delivery/service/ProductServiceTest.kt:
```diff
@@ -0,0 +1,81 @@
+
+
+    @Test
+    @DisplayName("Add valid product")
```
```suggestion
```

why do we need `@DisplayName`? The name of the test function is pretty clear.

In most cases `@DisplayName` is used in Java, because in java we don't have function names like
```kotlin
fun `some function name`()
```

ajax-torbin-v committed:
Resolved all issues

ajax-torbin-v committed:
Deleted redurdant test

ajax-gnatiuk-s commented on src/main/kotlin/com/example/delivery/controller/OrderController.kt:
```diff
@@ -0,0 +1,49 @@
+@RequestMapping("/orders")
+class OrderController(private val orderService: OrderService) {
+
+    @LogInvoke
```
<img width="575" alt="image" src="https://github.com/user-attachments/assets/be66444b-5e8a-4945-9ee5-cf7df06f3797">

Compile issues

ajax-torbin-v committed:
Deleted non-working annotation

ajax-shulepa-a commented on src/main/kotlin/com/example/delivery/config/GlobalExceptionHandler.kt:
```diff
@@ -0,0 +1,25 @@
+    @ExceptionHandler(NotFoundException::class)
+    fun handleNotFoundException(ex: NotFoundException, request: WebRequest): ResponseEntity<ErrorDetails> {
+        val errorDetails = ErrorDetails(
+            message = ex.message!!,
```
error may not contain a message. e.g. NullPointerException

ajax-shulepa-a commented on src/main/kotlin/com/example/delivery/mapper/OrderMapper.kt:
```diff
@@ -0,0 +1,49 @@
+    )
+
+    fun DomainOrder.toDTO(): OrderDTO = OrderDTO(
+        (id ?: "none").toString(),
```
```suggestion
        id!!.toHexString(),
```

ajax-shulepa-a commented on src/main/kotlin/com/example/delivery/domain/DomainOrder.kt:
```diff
@@ -0,0 +1,14 @@
+import java.math.BigDecimal
+
+data class DomainOrder(
+    val id: ObjectId? = null,
```
can your domain model exist without an id? 
```suggestion
    val id: ObjectId,
```

ajax-shulepa-a commented on src/main/kotlin/com/example/delivery/service/OrderService.kt:
```diff
@@ -0,0 +1,63 @@
+            totalPrice = totalPrice
+        )
+
+        val savedOrder: MongoOrder = orderRepository.save(order)
+        return savedOrder.toDomain()
```
```suggestion
        return orderRepository.save(order).toDomain()
```

ajax-shulepa-a commented on src/main/kotlin/com/example/delivery/service/UserService.kt:
```diff
@@ -0,0 +1,38 @@
+
+    fun add(createUserDTO: CreateUserDTO): DomainUser = userRepository.save(createUserDTO.toMongo()).toDomain()
+
+    fun getById(id: String): DomainUser {
+        val user: MongoUser = userRepository.findById(id)
+            ?: throw NotFoundException("User with id $id doesn't exists")
+        return user.toDomain()
+    }
```
```suggestion
    fun getById(id: String): DomainUser =
        userRepository.findById(id)?.toDomain()
            ?: throw NotFoundException("User with id $id doesn't exists")
    }
```

ajax-shulepa-a commented on src/test/kotlin/com/example/delivery/controller/OrderControllerTest.kt:
```diff
@@ -0,0 +1,118 @@
+    @Test
+    fun `should add order and return status is created`() {
+        //GIVEN
+        Mockito.`when`(orderService.add(createOrderDTO)).thenReturn(domainOrder)
```
in future, use mockito-kotlin or mockk, please
`when` looks ugly in kotlin :)

ajax-shulepa-a commented on src/testFixtures/kotlin/com/example/delivery/OrderFixture.kt:
```diff
@@ -0,0 +1,66 @@
+import java.math.BigDecimal
+
+object OrderFixture {
+    val order = MongoOrder(
+        id = ObjectId("123456789011121314151617"),
```
it is good practice to generate random values for test data. Follow it in next fixtures, please

ajax-torbin-v committed:
Reworked Domain model id's
Reworked mappers
